letsencrypt-win-simple/Scripts/ImportExchange.ps1

```diff
@@ -1,56 +1,95 @@
+Write-Debug -Message ('ExchangeServices: {0}' -f $ExchangeServices)
+Write-Debug -Message ('TargetHost: {0}' -f $TargetHost)
+Write-Debug -Message ('StorePath: {0}' -f $StorePath)
+
+
 ## Imports new cert thumbprint into the Exchange roles defined in the $ExchangeServices comma-separate list
 ## example: $ExchangeServices = IMAP,POP,IIS
 ## That will import the certificate for IMAP, POP, and IIS roles in exchange
 ## Valid services:  IMAP | POP | UM | IIS | SMTP | Federation | UMCallRouter
 
-## THIS SCRIPT IS INCOMPLETE AND UNTESTED
+## THIS SCRIPT IS INCOMPLETE AND *mostly* UNTESTED (some modifications have come in from people using it successfully)
 ## Documentation referenced from https://technet.microsoft.com/en-us/library/aa997231(v=exchg.160).aspx
 
 # Should work with exchange 2007 and higher
-Get-PSSnapin -registered | ? {$_.Name -match "Microsoft.Exchange.Management.PowerShell" -and ($_.Name -match "Admin" -or $_.Name -match "E2010" -or $_.Name -match "SnapIn")} | Add-PSSnapin -ErrorAction SilentlyContinue
+Get-PSSnapin -registered | Where-Object {$_.Name -match "Microsoft.Exchange.Management.PowerShell" -and ($_.Name -match "Admin" -or $_.Name -match "E2010" -or $_.Name -match "SnapIn")} | Add-PSSnapin -ErrorAction SilentlyContinue
 
 #$OldThumbprint = (Get-Item -Path RDS:\GatewayServer\SSLCertificate\Thumbprint).CurrentValue
 $CertInStore = Get-ChildItem -Path Cert:\LocalMachine -Recurse | Where-Object {$_.thumbprint -eq $NewCertThumbprint} | Sort-Object -Descending | Select-Object -f 1
-if($CertInStore){
-    try{
-        # Cert must exist in the personal store of machine to bind to RD Gateway
-        if($CertInStore.PSPath -notlike "*LocalMachine\My\*"){
-            $SourceStoreScope = 'LocalMachine'
-            $SourceStorename = $CertInStore.PSParentPath.split("\")[-1]
+try{
+    if($CertInStore.PSPath -notlike "*LocalMachine\My\*"){
+        "Cert thumbprint not found in the cert store... which means we should load it. This means TargetHost and StorePath must be specified"
+        
+        "Try to load certificate from store"
+        $importExchangeCertificateParameters = @{
+        FileName = (Join-Path -Path $StorePath -ChildPath "$TargetHost.pfx")
+        FriendlyName = $TargetHost
+        PrivateKeyExportable = $true
+        }
+        
+        $null = Import-ExchangeCertificate @importExchangeCertificateParameters -ErrorAction Stop
 
-            $SourceStore = New-Object  -TypeName System.Security.Cryptography.X509Certificates.X509Store  -ArgumentList $SourceStorename, $SourceStoreScope
-            $SourceStore.Open([System.Security.Cryptography.X509Certificates.OpenFlags]::ReadOnly)
 
-            $cert = $SourceStore.Certificates | Where-Object {$_.thumbprint -eq $CertInStore.Thumbprint}
+        # Now try to find it again...
+        $CertInStore = Get-ChildItem -Path Cert:\LocalMachine -Recurse | Where-Object {$_.thumbprint -eq $NewCertThumbprint} | Sort-Object -Descending | Select-Object -f 1
```

Comment:
We should not need to do this, `Import-ExchangeCertificate` returns the certificate if you add `$CertInStore = Import-ExchangeCertificate @importExchangeCertificateParameters -ErrorAction Stop` or throws an error (stops) if it can't find the certificate.

If you keep this row, the same comment about `-Descending` as above.

Reply:
Seems cleaner to do this so we are sure to get the right type of object. I don't have an exchange install to know if the passthrough from `import-exchangecertificate` returns the same type of object as `get-childitem`. I like to play it safe in that regard.